text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+this, which when applied like so:
+
+```rust
+#[wasm_bindgen(method, structural)]
+fn method(this: &Parent);
+```
+
+means that the following JS code is generated:
+
+```js
+const Parent_method_target = function() { this.method(); };
+
+// ...
+```
+
+Here we can see that a JS function shim is generated instead of using the raw
+function value in the prototype. This, however, means that our example above
+will indeed print `parent` and then `child` because JS is using prototype
+lookups to find the `method` method.
+
+Phew! Ok with all that information, we can see that **if `structural` is omitted
+then JS class hierarchies can be subtly incorrect when methods taking parent
+classes are passed child classes which override methods**.
+
+An easy solution to this problem is to simply use `structural` everywhere, so...
+let's propose that! Consequently, this RFC proposes changing `#[wasm_bindgen]`
+to act as if all bindings are labeled as `structural`. This will not be a
+breaking change because the generated bindings will still have the same behavior
+as before, they'll just handle subclassing correctly!
+
+### Adding `#[wasm_bindgen(final)]`
+
+Since `structural` is not the default today we don't actually have a name for
+the default behavior of `#[wasm_bindgen]` today. This RFC proposes adding a new
+attribute to `#[wasm_bindgen]`, `final`, which indicates that it should have
+today's behavior.
+
+When attached to an attribute or method, the `final` attribute indicates that
+the method or attribute should be processed through the `prototype` of a class
+rather than looked up structurally via the prototype chain.
+
+You can think of this as "everything today is `final` by default".
+
+### Why is it ok to make `structural` the default?
+
+One pretty reasonable question you might have at this point is "why, if
+`structural` is the default today, is it ok to switch?" To answer this, let's
+first explore why `final` is the default today!
+
+From its inception `wasm-bindgen` has been designed with the future [host
+bindings] proposal for WebAssembly. The host bindings proposal promises
+faster-than-JS DOM access by removing many of the dynamic checks necessary when
+calling DOM methods. This proposal, however, is still in relatively early stages
+and hasn't been implemented in any browser yet (as far as we know).
+
+In WebAssembly on the web all imported functions must be plain old JS functions.
+They're all currently invoked with `undefined` as the `this` parameter. With
+host bindings, however, there's a way to say that an imported function uses the
+first argument to the function as the `this` parameter (like `Function.call` in
+JS). This in turn brings the promise of *eliminating any shim functions
+necessary when calling imported functionality*.
+
+As an example, today for `#[wasm_bindgen(method)] fn parent(this: &Parent);` we
+generate JS that looks like:
+
+```rust
+#[wasm_bindgen(method, structural)]
```

Comment:
Is there a reason this is marked as `structural` even though above it doesn't use `structural`?

Reply:
This was a mistake